montara/des_tile.py

```diff
@@ -674,6 +674,8 @@ def setup(self, config, base, file_num, logger):
                 base['gal']['rng_num'] = 1
             if 'star' in base:
                 base['star']['rng_num'] = 1
+            if 'psf' in base:
+                base['psf']['rng_num'] = 1
```

Comment:
That doesn't seem right to me.  It looks like rng_num=0 is used for things that are supposed to change on every exposure.  The second one (rng_num=1) is for things that you want to repeat with the same random values for each exposure in a given tile.

I would have thought you'd want the PSF to change every exposure, so it would use rng_num=0.

Reply:
We are using the color of the galaxy to evaluate the PSF model, so it has to change with the galaxy, not randomly on its own.

Reply:
It would be helpful for me to explain a bit more to see what you think. The current galsim config we have is (just a snippet):

```yaml
psf:
  type: DES_Piff
  depixelize: false
  gi_color: { type: Eval, str: "mag_g - mag_i", fmag_g: { type: catalog_sampler_value, col: mag_g }, fmag_i: { type: catalog_sampler_value, col: mag_i } }
  iz_color: { type: Eval, str: "mag_i - mag_z", fmag_i: { type: catalog_sampler_value, col: mag_i }, fmag_z: { type: catalog_sampler_value, col: mag_z } }

#No knots
gal:
    type: Sum
    items:
        - type: Exponential
          half_light_radius: { type: catalog_sampler_value, col: bdf_hlr }
          flux: { type: Eval, str: "1-fracdev", ffracdev: { type: catalog_sampler_value, col: bdf_fracdev } }

        - type: DeVaucouleurs
          half_light_radius: '@gal.items.0.half_light_radius'
          flux: "$1-@gal.items.0.flux"

    ellip:
        type: GBeta
        g: { type: Eval, str: "np.sqrt(g1**2 + g2**2)", fg1: { type: catalog_sampler_value, col: bdf_g1 }, fg2: { type: catalog_sampler_value, col: bdf_g2 } }
        beta: { type: Random }

    rotate: {type: Random}

    flux:
        type: Eval
        str: "10**(0.4*(mag_zp-mag))"
        fmag: { type: catalog_sampler_value, col: *gal_mag_col }

    gsparams:
        folding_threshold: 1.0e-4
```

What we see in the currently buggy sims is that the magnitudes in each band are effectively randomized when we include calls to the `catalog_sampler_value` type in the PSF section. If the PSF section uses a different RNG than the galaxy one, then we think two things happen:

1. the wrong object is drawn from the catalog since the RNG used is not the same as the galaxy one
2. somehow, and on this we are not sure, by having the `catalog_sampler_value` type tied to two different RNGs, things are getting out of sync across the different bands

The `catalog_sampler_value` has a single piece of global state in the base config (https://github.com/des-science/montara/blob/main/montara/catalogsampler.py#L116) and we suspect that is getting messed up by using it two different RNGs.

Reply:
Sounds like you should put `rng_num: 1` directly in the psf.gi_color and psf.iz_color fields.  Not in the overall psf field.  If there are no other bits of randomness in the psf computation, that may end up functionally the same as what this PR does, but it would be logically correct, which this proposal is not.

Reply:
I put this special case into the code directly so that the config files don't have to be synced up with the knowledge of which rng_num is being used for the galaxies.

Reply:
Well, it's up to you I guess.  But if there is anything else that is random about the PSF (e.g. which piff file is getting used), then this will do the wrong thing.  You want everything about the psf to use rng_num 0 *except* the bits that are actually about the galaxy.  
Even if there is nothing else random now, I would still prefer to be explicit in the config file rather than try to be clever in the program and modify the config file behind the user's back.  Like I would probably recommend not doing any of these rng_num=1 bits in the code, and just document that this output type automatically makes an additional rng, which repeats at the tile cadence, so that should be used for any field that wants to have the same values when you're at the same place on the sky.  Then let the user put rng_num=1 at the right places.  (Such as the PSF color items.)

Reply:
Yeah I see your point and tend to agree. Unfortunately, the explicit approach was already not chosen by the code as originally developed. Switching to it now creates more issues than it solves because then we have logic for a single thing (rng nums) spread across more than one place. I think pilling on more magic on top of the magic we already inherited is the least worst of the options.

Reply:
My opinion is that being explicit here for the colors is the least worst, but YMMV I suppose.

Reply:
One thing to note: the most recent version of the config that uses PIFF looks like
```
psf:                                                                                                                                                                                                                                           
  type: DES_Piff
  depixelize: false
  gi_color:
    type: Eval
    str: "mag_g_star - mag_i_star if object_type_list[@obj_num-start_obj_num]=='star' else mag_g_gal - mag_i_gal"
    fmag_g_gal: { type: catalog_sampler_value, col: mag_g_dered }
    fmag_i_gal: { type: catalog_sampler_value, col: mag_i_dered }
    fmag_g_star: { type: DESStarValue, col: "gmag" }
    fmag_i_star: { type: DESStarValue, col: "imag" }
  iz_color:
    type: Eval
    str: "mag_i_star - mag_z_star if object_type_list[@obj_num-start_obj_num]=='star' else mag_i_gal - mag_z_gal"
    fmag_i_gal: { type: catalog_sampler_value, col: mag_i_dered }
    fmag_z_gal: { type: catalog_sampler_value, col: mag_z_dered }
    fmag_i_star: { type: DESStarValue, col: "imag" }
    fmag_z_star: { type: DESStarValue, col: "zmag" }
```
because PIFF needs to know the color of the object being rendered, whether it is a star or a galaxy. I don't think this has any bearing on the above conversation, but it is good to keep in mind that we want to be syncing the PSF color rng with _both_ that of the stars and the galaxies

Reply:
another solution would be to somehow propagate each objects color (e.g., as an attribute of the gsobject), but I don't know of a way to accomplish this

Reply:
another solution would be to somehow propagate each objects color (e.g., as an attribute of the gsobject), but I don't know of a way to accomplish this

Reply:
I think I have the galaxies fixed. I'll fix the stars too.